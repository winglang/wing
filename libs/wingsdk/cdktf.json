{
  "language": "typescript",
  "app": "echo noop",
  "terraformProviders": [
    "aws@~>5.31.0",
    "random@~>3.5.1",
    "azurerm@~>3.54.0",
    "google@~>4.63.1"
  ],
  "codeMakerOutput": "src/.gen",
  "projectId": "93afdbfa-23ed-40cf-9ce4-495b3289c519",
  "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run \"npx projen\"."
}
