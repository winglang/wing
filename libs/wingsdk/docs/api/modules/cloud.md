[@monadahq/wingsdk](../README.md) / [Exports](../modules.md) / cloud

# Namespace: cloud

## Table of contents

### Enumerations

- [BucketInflightMethods](../enums/cloud.BucketInflightMethods.md)
- [FunctionInflightMethods](../enums/cloud.FunctionInflightMethods.md)
- [QueueInflightMethods](../enums/cloud.QueueInflightMethods.md)

### Classes

- [Bucket](../classes/cloud.Bucket.md)
- [BucketBase](../classes/cloud.BucketBase.md)
- [Function](../classes/cloud.Function.md)
- [FunctionBase](../classes/cloud.FunctionBase.md)
- [Queue](../classes/cloud.Queue.md)
- [QueueBase](../classes/cloud.QueueBase.md)
- [Resource](../classes/cloud.Resource.md)

### Interfaces

- [BucketProps](../interfaces/cloud.BucketProps.md)
- [FunctionProps](../interfaces/cloud.FunctionProps.md)
- [IBucketClient](../interfaces/cloud.IBucketClient.md)
- [IFunctionClient](../interfaces/cloud.IFunctionClient.md)
- [IQueueClient](../interfaces/cloud.IQueueClient.md)
- [QueueOnMessageProps](../interfaces/cloud.QueueOnMessageProps.md)
- [QueueProps](../interfaces/cloud.QueueProps.md)

### Variables

- [BUCKET\_TYPE](cloud.md#bucket_type)
- [FUNCTION\_TYPE](cloud.md#function_type)
- [QUEUE\_TYPE](cloud.md#queue_type)

## Variables

### BUCKET\_TYPE

• `Const` **BUCKET\_TYPE**: ``"wingsdk.cloud.Bucket"``

Global identifier for `Bucket`.

#### Defined in

[src/cloud/bucket.ts:9](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/bucket.ts#L9)

___

### FUNCTION\_TYPE

• `Const` **FUNCTION\_TYPE**: ``"wingsdk.cloud.Function"``

Global identifier for `Function`.

#### Defined in

[src/cloud/function.ts:9](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/function.ts#L9)

___

### QUEUE\_TYPE

• `Const` **QUEUE\_TYPE**: ``"wingsdk.cloud.Queue"``

Global identifier for `Queue`.

#### Defined in

[src/cloud/queue.ts:10](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/queue.ts#L10)
