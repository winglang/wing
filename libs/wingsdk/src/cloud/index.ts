export * from "./api";
export * from "./bucket";
export * from "./counter";
export * from "./domain";
export * from "./endpoint";
export * from "./function";
export * from "./on-deploy";
export * from "./queue";
export * from "./schedule";
export * from "./secret";
export * from "./service";
export * from "./topic";
export * from "./website";
