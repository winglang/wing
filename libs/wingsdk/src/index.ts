// this file must only export the standard library types and not target-specific types.
export * as cloud from "./cloud";
export * as core from "./core";
export * as ex from "./ex";
export * as expect from "./expect";
export * as fs from "./fs";
export * as helpers from "./helpers";
export * as http from "./http";
export * as math from "./math";
export * as platform from "./platform";
export * as regex from "./regex";
export * as aws from "./shared-aws";
export * as simulator from "./simulator";
export * as std from "./std";
export * as sim from "./target-sim";
export * as ui from "./ui";
export * as util from "./util";
