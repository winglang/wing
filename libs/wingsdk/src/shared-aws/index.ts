export * from "./api";
export * from "./bucket";
export * from "./counter";
export * from "./domain";
export * from "./function";
export * from "./queue";
export * from "./secret";
export * from "./table";
export * from "./topic";
export * from "./types";
export * from "./website";
