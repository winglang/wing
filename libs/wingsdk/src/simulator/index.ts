export * from "./client";
export * from "./simulator";
export * from "./tree";
