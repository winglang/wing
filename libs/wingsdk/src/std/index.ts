export * from "./array";
export * from "./bool";
export * from "./duration";
export * from "./json";
export * from "./map";
export * from "./number";
export * from "./range";
export * from "./resource";
export * from "./set";
export * from "./string";
export * from "./test";
export * from "./util";
