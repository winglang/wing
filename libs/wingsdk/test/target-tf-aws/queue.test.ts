import * as cdktf from "cdktf";
import { test, expect } from "vitest";
import { Function, IFunctionClient, Queue } from "../../src/cloud";
import { QueueRef } from "../../src/shared-aws";
import { Testing } from "../../src/simulator";
import * as std from "../../src/std";
import * as tfaws from "../../src/target-tf-aws";
import { SimApp } from "../sim-app";
import {
  mkdtemp,
  sanitizeCode,
  tfResourcesOf,
  tfSanitize,
  treeJsonOf,
} from "../util";

test("default queue behavior", () => {
  // GIVEN
  const app = new tfaws.App({ outdir: mkdtemp(), entrypointDir: __dirname });
  new Queue(app, "Queue");
  const output = app.synth();

  // THEN
  expect(tfResourcesOf(output)).toEqual(["aws_sqs_queue"]);
  expect(tfSanitize(output)).toMatchSnapshot();
  expect(treeJsonOf(app.outdir)).toMatchSnapshot();
});

test("queue with custom timeout", () => {
  // GIVEN
  const app = new tfaws.App({ outdir: mkdtemp(), entrypointDir: __dirname });
  new Queue(app, "Queue", {
    timeout: std.Duration.fromSeconds(30),
  });
  const output = app.synth();

  // THEN
  expect(tfResourcesOf(output)).toEqual(["aws_sqs_queue"]);
  expect(tfSanitize(output)).toMatchSnapshot();
  expect(treeJsonOf(app.outdir)).toMatchSnapshot();
});

test("queue with custom retention", () => {
  // GIVEN
  const app = new tfaws.App({ outdir: mkdtemp(), entrypointDir: __dirname });
  new Queue(app, "Queue", {
    retentionPeriod: std.Duration.fromMinutes(30),
  });
  const output = app.synth();

  // THEN
  expect(tfResourcesOf(output)).toEqual(["aws_sqs_queue"]);
  expect(tfSanitize(output)).toMatchSnapshot();
  expect(treeJsonOf(app.outdir)).toMatchSnapshot();
});

test("queue with a consumer function", () => {
  // GIVEN
  const app = new tfaws.App({ outdir: mkdtemp(), entrypointDir: __dirname });
  const queue = new Queue(app, "Queue", {
    timeout: std.Duration.fromSeconds(30),
  });
  const processor = Testing.makeHandler(`\
async handle(event) {
  console.log("Received " + event.name);
}`);
  const processorFn = queue.setConsumer(processor);
  const output = app.synth();

  // THEN
  expect(sanitizeCode(processorFn._toInflight())).toMatchSnapshot();

  expect(tfResourcesOf(output)).toEqual([
    "aws_cloudwatch_log_group", // log group for function
    "aws_iam_role", // role for function
    "aws_iam_role_policy", // policy for role
    "aws_iam_role_policy_attachment", // execution policy for role
    "aws_lambda_event_source_mapping", // connection between queue and function
    "aws_lambda_function", // processor function
    "aws_s3_bucket", // S3 bucket for code
    "aws_s3_object", // S3 object for code
    "aws_sqs_queue", // main queue
  ]);
  expect(tfSanitize(output)).toMatchSnapshot();
  expect(treeJsonOf(app.outdir)).toMatchSnapshot();
});

test("queue name valid", () => {
  // GIVEN
  const app = new tfaws.App({ outdir: mkdtemp(), entrypointDir: __dirname });
  const queue = new Queue(app, "The-Incredible_Queue-01");
  const output = app.synth();

  // THEN
  expect(
    cdktf.Testing.toHaveResourceWithProperties(output, "aws_sqs_queue", {
      name: `The-Incredible_Queue-01-${queue.node.addr.substring(0, 8)}`,
    })
  ).toBeTruthy();
  expect(tfSanitize(output)).toMatchSnapshot();
  expect(treeJsonOf(app.outdir)).toMatchSnapshot();
});

test("replace invalid character from queue name", () => {
  // GIVEN
  const app = new tfaws.App({ outdir: mkdtemp(), entrypointDir: __dirname });
  const queue = new Queue(app, "The*Incredible$Queue");
  const output = app.synth();

  // THEN
  expect(
    cdktf.Testing.toHaveResourceWithProperties(output, "aws_sqs_queue", {
      name: `The-Incredible-Queue-${queue.node.addr.substring(0, 8)}`,
    })
  ).toBeTruthy();
  expect(tfSanitize(output)).toMatchSnapshot();
  expect(treeJsonOf(app.outdir)).toMatchSnapshot();
});

test("QueueRef fails with an invalid ARN", async () => {
  const app = new SimApp();
  expect(() => {
    new QueueRef(app, "QueueRef", "not-an_arn");
  }).toThrow(/"not-an_arn" is not a valid Amazon SQS ARN/);
});

test("QueueRef in a SimApp can be used to reference an existing queue within a simulated app", async () => {
  const app = new SimApp();

  const queueRef = new QueueRef(
    app,
    "QueueRef",
    "arn:aws:sqs:us-west-2:123456789012:MyQueue1234"
  );

  // we can't really make a remote call here, so we'll just check that
  // we have the right inflight client with the right queue name.
  new Function(
    app,
    "Function",
    Testing.makeHandler(
      `async handle() {
        if (!this.queue.client) {
          throw new Error("Queue AWS SDK client not found");
        }
        if (this.queue.client.constructor.name !== "SQSClient") {
          throw new Error("Queue AWS SDK client is not an SQSClient");
        }
        return this.queue._queueUrlOrArn; // yes, internal stuff
      }`,
      {
        queue: { obj: queueRef, ops: ["push"] },
      }
    )
  );

  expect(queueRef.queueArn).toStrictEqual(
    "arn:aws:sqs:us-west-2:123456789012:MyQueue1234"
  );

  const sim = await app.startSimulator();

  const fn = sim.getResource("root/Function") as IFunctionClient;

  const reply = await fn.invoke();
  expect(reply).toStrictEqual("arn:aws:sqs:us-west-2:123456789012:MyQueue1234");
});

test("QueueRef in an TFAWS app can be used to reference an existing queue", () => {
  // GIVEN
  const app = new tfaws.App({ outdir: mkdtemp(), entrypointDir: __dirname });
  const queue = new QueueRef(
    app,
    "QueueRef",
    "arn:aws:sqs:us-west-2:123456789012:MyQueue1234"
  );

  const handler = Testing.makeHandler(``, {
    queue: { obj: queue, ops: ["push", "approxSize"] },
  });

  new Function(app, "Function", handler);

  const output = app.synth();

  // THEN

  const statements = JSON.parse(
    JSON.parse(output).resource.aws_iam_role_policy
      .Function_IamRolePolicy_E3B26607.policy
  ).Statement;

  expect(statements).toStrictEqual([
    {
      Action: ["sqs:GetQueueUrl"],
      Effect: "Allow",
      Resource: ["arn:aws:sqs:us-west-2:123456789012:MyQueue1234"],
    },
    {
      Action: ["sqs:SendMessage"],
      Effect: "Allow",
      Resource: ["arn:aws:sqs:us-west-2:123456789012:MyQueue1234"],
    },
    {
      Action: ["sqs:GetQueueAttributes"],
      Effect: "Allow",
      Resource: ["arn:aws:sqs:us-west-2:123456789012:MyQueue1234"],
    },
  ]);

  expect(tfSanitize(output)).toMatchSnapshot();
  expect(treeJsonOf(app.outdir)).toMatchSnapshot();
});
