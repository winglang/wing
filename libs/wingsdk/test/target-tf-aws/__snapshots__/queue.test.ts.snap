// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`default queue behavior 1`] = `
{
  "resource": {
    "aws_sqs_queue": {
      "Queue": {
        "message_retention_seconds": 3600,
        "name": "Queue-c822c726",
        "visibility_timeout_seconds": 30,
      },
    },
  },
}
`;

exports[`default queue behavior 2`] = `
{
  "tree": {
    "children": {
      "root": {
        "children": {
          "Default": {
            "children": {
              "Queue": {
                "children": {
                  "Default": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformResource",
                      "version": "0.20.3",
                    },
                    "id": "Default",
                    "path": "root/Default/Queue/Default",
                  },
                },
                "constructInfo": {
                  "fqn": "constructs.Construct",
                  "version": "10.3.0",
                },
                "display": {
                  "description": "A distributed message queue",
                  "title": "Queue",
                },
                "id": "Queue",
                "path": "root/Default/Queue",
              },
              "aws": {
                "constructInfo": {
                  "fqn": "cdktf.TerraformProvider",
                  "version": "0.20.3",
                },
                "id": "aws",
                "path": "root/Default/aws",
              },
            },
            "constructInfo": {
              "fqn": "constructs.Construct",
              "version": "10.3.0",
            },
            "id": "Default",
            "path": "root/Default",
          },
          "backend": {
            "constructInfo": {
              "fqn": "cdktf.LocalBackend",
              "version": "0.20.3",
            },
            "id": "backend",
            "path": "root/backend",
          },
        },
        "constructInfo": {
          "fqn": "cdktf.TerraformStack",
          "version": "0.20.3",
        },
        "id": "root",
        "path": "root",
      },
    },
    "constructInfo": {
      "fqn": "cdktf.App",
      "version": "0.20.3",
    },
    "id": "App",
    "path": "",
  },
  "version": "tree-0.1",
}
`;

exports[`queue name valid 1`] = `
{
  "resource": {
    "aws_sqs_queue": {
      "The-Incredible_Queue-01": {
        "message_retention_seconds": 3600,
        "name": "The-Incredible_Queue-01-c8d1a14d",
        "visibility_timeout_seconds": 30,
      },
    },
  },
}
`;

exports[`queue name valid 2`] = `
{
  "tree": {
    "children": {
      "root": {
        "children": {
          "Default": {
            "children": {
              "The-Incredible_Queue-01": {
                "children": {
                  "Default": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformResource",
                      "version": "0.20.3",
                    },
                    "id": "Default",
                    "path": "root/Default/The-Incredible_Queue-01/Default",
                  },
                },
                "constructInfo": {
                  "fqn": "constructs.Construct",
                  "version": "10.3.0",
                },
                "display": {
                  "description": "A distributed message queue",
                  "title": "Queue",
                },
                "id": "The-Incredible_Queue-01",
                "path": "root/Default/The-Incredible_Queue-01",
              },
              "aws": {
                "constructInfo": {
                  "fqn": "cdktf.TerraformProvider",
                  "version": "0.20.3",
                },
                "id": "aws",
                "path": "root/Default/aws",
              },
            },
            "constructInfo": {
              "fqn": "constructs.Construct",
              "version": "10.3.0",
            },
            "id": "Default",
            "path": "root/Default",
          },
          "backend": {
            "constructInfo": {
              "fqn": "cdktf.LocalBackend",
              "version": "0.20.3",
            },
            "id": "backend",
            "path": "root/backend",
          },
        },
        "constructInfo": {
          "fqn": "cdktf.TerraformStack",
          "version": "0.20.3",
        },
        "id": "root",
        "path": "root",
      },
    },
    "constructInfo": {
      "fqn": "cdktf.App",
      "version": "0.20.3",
    },
    "id": "App",
    "path": "",
  },
  "version": "tree-0.1",
}
`;

exports[`queue with a consumer function 1`] = `"new (require(\\"[REDACTED]/wingsdk/src/shared-aws/function.inflight\\")).FunctionClient(process.env[\\"FUNCTION_NAME_fd688af4\\"], \\"root/Default/Queue-SetConsumer0\\")"`;

exports[`queue with a consumer function 2`] = `
{
  "resource": {
    "aws_cloudwatch_log_group": {
      "Queue-SetConsumer0_CloudwatchLogGroup_56C2891C": {
        "name": "/aws/lambda/Queue-SetConsumer0-c83c303c",
        "retention_in_days": 30,
      },
    },
    "aws_iam_role": {
      "Queue-SetConsumer0_IamRole_7F9ED9ED": {
        "assume_role_policy": "{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":[{\\"Action\\":\\"sts:AssumeRole\\",\\"Principal\\":{\\"Service\\":\\"lambda.amazonaws.com\\"},\\"Effect\\":\\"Allow\\"}]}",
      },
    },
    "aws_iam_role_policy": {
      "Queue-SetConsumer0_IamRolePolicy_0299B5AB": {
        "policy": "{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":[{\\"Action\\":[\\"sqs:ReceiveMessage\\",\\"sqs:ChangeMessageVisibility\\",\\"sqs:GetQueueUrl\\",\\"sqs:DeleteMessage\\",\\"sqs:GetQueueAttributes\\"],\\"Resource\\":[\\"\${aws_sqs_queue.Queue.arn}\\"],\\"Effect\\":\\"Allow\\"}]}",
        "role": "\${aws_iam_role.Queue-SetConsumer0_IamRole_7F9ED9ED.name}",
      },
    },
    "aws_iam_role_policy_attachment": {
      "Queue-SetConsumer0_IamRolePolicyAttachment_4A4C5C5D": {
        "policy_arn": "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
        "role": "\${aws_iam_role.Queue-SetConsumer0_IamRole_7F9ED9ED.name}",
      },
    },
    "aws_lambda_event_source_mapping": {
      "Queue_EventSourceMapping_8332F7DC": {
        "batch_size": 1,
        "event_source_arn": "\${aws_sqs_queue.Queue.arn}",
        "function_name": "\${aws_lambda_function.Queue-SetConsumer0.function_name}",
      },
    },
    "aws_lambda_function": {
      "Queue-SetConsumer0": {
        "architectures": [
          "arm64",
        ],
        "environment": {
          "variables": {
            "NODE_OPTIONS": "--enable-source-maps",
            "WING_FUNCTION_NAME": "Queue-SetConsumer0-c83c303c",
          },
        },
        "function_name": "Queue-SetConsumer0-c83c303c",
        "handler": "index.handler",
        "memory_size": 1024,
        "publish": true,
        "role": "\${aws_iam_role.Queue-SetConsumer0_IamRole_7F9ED9ED.arn}",
        "runtime": "nodejs20.x",
        "s3_bucket": "\${aws_s3_bucket.Code.bucket}",
        "s3_key": "\${aws_s3_object.Queue-SetConsumer0_S3Object_2AD0A795.key}",
        "timeout": "\${aws_sqs_queue.Queue.visibility_timeout_seconds}",
        "vpc_config": {
          "security_group_ids": [],
          "subnet_ids": [],
        },
      },
    },
    "aws_s3_bucket": {
      "Code": {
        "bucket_prefix": "code-c84a50b1-",
      },
    },
    "aws_s3_object": {
      "Queue-SetConsumer0_S3Object_2AD0A795": {
        "bucket": "\${aws_s3_bucket.Code.bucket}",
        "key": "<key>",
        "source": "<source>",
      },
    },
    "aws_sqs_queue": {
      "Queue": {
        "message_retention_seconds": 3600,
        "name": "Queue-c822c726",
        "visibility_timeout_seconds": 30,
      },
    },
  },
}
`;

exports[`queue with a consumer function 3`] = `
{
  "tree": {
    "children": {
      "root": {
        "children": {
          "Default": {
            "children": {
              "Code": {
                "constructInfo": {
                  "fqn": "cdktf.TerraformResource",
                  "version": "0.20.3",
                },
                "id": "Code",
                "path": "root/Default/Code",
              },
              "Queue": {
                "children": {
                  "Default": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformResource",
                      "version": "0.20.3",
                    },
                    "id": "Default",
                    "path": "root/Default/Queue/Default",
                  },
                  "EventSourceMapping": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformResource",
                      "version": "0.20.3",
                    },
                    "id": "EventSourceMapping",
                    "path": "root/Default/Queue/EventSourceMapping",
                  },
                },
                "constructInfo": {
                  "fqn": "constructs.Construct",
                  "version": "10.3.0",
                },
                "display": {
                  "description": "A distributed message queue",
                  "title": "Queue",
                },
                "id": "Queue",
                "path": "root/Default/Queue",
              },
              "Queue-SetConsumer0": {
                "children": {
                  "Asset": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformAsset",
                      "version": "0.20.3",
                    },
                    "id": "Asset",
                    "path": "root/Default/Queue-SetConsumer0/Asset",
                  },
                  "CloudwatchLogGroup": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformResource",
                      "version": "0.20.3",
                    },
                    "id": "CloudwatchLogGroup",
                    "path": "root/Default/Queue-SetConsumer0/CloudwatchLogGroup",
                  },
                  "Default": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformResource",
                      "version": "0.20.3",
                    },
                    "id": "Default",
                    "path": "root/Default/Queue-SetConsumer0/Default",
                  },
                  "IamRole": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformResource",
                      "version": "0.20.3",
                    },
                    "id": "IamRole",
                    "path": "root/Default/Queue-SetConsumer0/IamRole",
                  },
                  "IamRolePolicy": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformResource",
                      "version": "0.20.3",
                    },
                    "id": "IamRolePolicy",
                    "path": "root/Default/Queue-SetConsumer0/IamRolePolicy",
                  },
                  "IamRolePolicyAttachment": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformResource",
                      "version": "0.20.3",
                    },
                    "id": "IamRolePolicyAttachment",
                    "path": "root/Default/Queue-SetConsumer0/IamRolePolicyAttachment",
                  },
                  "S3Object": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformResource",
                      "version": "0.20.3",
                    },
                    "id": "S3Object",
                    "path": "root/Default/Queue-SetConsumer0/S3Object",
                  },
                },
                "constructInfo": {
                  "fqn": "constructs.Construct",
                  "version": "10.3.0",
                },
                "display": {
                  "description": "A cloud function (FaaS)",
                  "title": "Function",
                },
                "id": "Queue-SetConsumer0",
                "path": "root/Default/Queue-SetConsumer0",
              },
              "aws": {
                "constructInfo": {
                  "fqn": "cdktf.TerraformProvider",
                  "version": "0.20.3",
                },
                "id": "aws",
                "path": "root/Default/aws",
              },
            },
            "constructInfo": {
              "fqn": "constructs.Construct",
              "version": "10.3.0",
            },
            "id": "Default",
            "path": "root/Default",
          },
          "backend": {
            "constructInfo": {
              "fqn": "cdktf.LocalBackend",
              "version": "0.20.3",
            },
            "id": "backend",
            "path": "root/backend",
          },
        },
        "constructInfo": {
          "fqn": "cdktf.TerraformStack",
          "version": "0.20.3",
        },
        "id": "root",
        "path": "root",
      },
    },
    "constructInfo": {
      "fqn": "cdktf.App",
      "version": "0.20.3",
    },
    "id": "App",
    "path": "",
  },
  "version": "tree-0.1",
}
`;

exports[`queue with custom retention 1`] = `
{
  "resource": {
    "aws_sqs_queue": {
      "Queue": {
        "message_retention_seconds": 1800,
        "name": "Queue-c822c726",
        "visibility_timeout_seconds": 30,
      },
    },
  },
}
`;

exports[`queue with custom retention 2`] = `
{
  "tree": {
    "children": {
      "root": {
        "children": {
          "Default": {
            "children": {
              "Queue": {
                "children": {
                  "Default": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformResource",
                      "version": "0.20.3",
                    },
                    "id": "Default",
                    "path": "root/Default/Queue/Default",
                  },
                },
                "constructInfo": {
                  "fqn": "constructs.Construct",
                  "version": "10.3.0",
                },
                "display": {
                  "description": "A distributed message queue",
                  "title": "Queue",
                },
                "id": "Queue",
                "path": "root/Default/Queue",
              },
              "aws": {
                "constructInfo": {
                  "fqn": "cdktf.TerraformProvider",
                  "version": "0.20.3",
                },
                "id": "aws",
                "path": "root/Default/aws",
              },
            },
            "constructInfo": {
              "fqn": "constructs.Construct",
              "version": "10.3.0",
            },
            "id": "Default",
            "path": "root/Default",
          },
          "backend": {
            "constructInfo": {
              "fqn": "cdktf.LocalBackend",
              "version": "0.20.3",
            },
            "id": "backend",
            "path": "root/backend",
          },
        },
        "constructInfo": {
          "fqn": "cdktf.TerraformStack",
          "version": "0.20.3",
        },
        "id": "root",
        "path": "root",
      },
    },
    "constructInfo": {
      "fqn": "cdktf.App",
      "version": "0.20.3",
    },
    "id": "App",
    "path": "",
  },
  "version": "tree-0.1",
}
`;

exports[`queue with custom timeout 1`] = `
{
  "resource": {
    "aws_sqs_queue": {
      "Queue": {
        "message_retention_seconds": 3600,
        "name": "Queue-c822c726",
        "visibility_timeout_seconds": 30,
      },
    },
  },
}
`;

exports[`queue with custom timeout 2`] = `
{
  "tree": {
    "children": {
      "root": {
        "children": {
          "Default": {
            "children": {
              "Queue": {
                "children": {
                  "Default": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformResource",
                      "version": "0.20.3",
                    },
                    "id": "Default",
                    "path": "root/Default/Queue/Default",
                  },
                },
                "constructInfo": {
                  "fqn": "constructs.Construct",
                  "version": "10.3.0",
                },
                "display": {
                  "description": "A distributed message queue",
                  "title": "Queue",
                },
                "id": "Queue",
                "path": "root/Default/Queue",
              },
              "aws": {
                "constructInfo": {
                  "fqn": "cdktf.TerraformProvider",
                  "version": "0.20.3",
                },
                "id": "aws",
                "path": "root/Default/aws",
              },
            },
            "constructInfo": {
              "fqn": "constructs.Construct",
              "version": "10.3.0",
            },
            "id": "Default",
            "path": "root/Default",
          },
          "backend": {
            "constructInfo": {
              "fqn": "cdktf.LocalBackend",
              "version": "0.20.3",
            },
            "id": "backend",
            "path": "root/backend",
          },
        },
        "constructInfo": {
          "fqn": "cdktf.TerraformStack",
          "version": "0.20.3",
        },
        "id": "root",
        "path": "root",
      },
    },
    "constructInfo": {
      "fqn": "cdktf.App",
      "version": "0.20.3",
    },
    "id": "App",
    "path": "",
  },
  "version": "tree-0.1",
}
`;

exports[`replace invalid character from queue name 1`] = `
{
  "resource": {
    "aws_sqs_queue": {
      "TheIncredibleQueue": {
        "message_retention_seconds": 3600,
        "name": "The-Incredible-Queue-c8305287",
        "visibility_timeout_seconds": 30,
      },
    },
  },
}
`;

exports[`replace invalid character from queue name 2`] = `
{
  "tree": {
    "children": {
      "root": {
        "children": {
          "Default": {
            "children": {
              "The*Incredible$Queue": {
                "children": {
                  "Default": {
                    "constructInfo": {
                      "fqn": "cdktf.TerraformResource",
                      "version": "0.20.3",
                    },
                    "id": "Default",
                    "path": "root/Default/The*Incredible$Queue/Default",
                  },
                },
                "constructInfo": {
                  "fqn": "constructs.Construct",
                  "version": "10.3.0",
                },
                "display": {
                  "description": "A distributed message queue",
                  "title": "Queue",
                },
                "id": "The*Incredible$Queue",
                "path": "root/Default/The*Incredible$Queue",
              },
              "aws": {
                "constructInfo": {
                  "fqn": "cdktf.TerraformProvider",
                  "version": "0.20.3",
                },
                "id": "aws",
                "path": "root/Default/aws",
              },
            },
            "constructInfo": {
              "fqn": "constructs.Construct",
              "version": "10.3.0",
            },
            "id": "Default",
            "path": "root/Default",
          },
          "backend": {
            "constructInfo": {
              "fqn": "cdktf.LocalBackend",
              "version": "0.20.3",
            },
            "id": "backend",
            "path": "root/backend",
          },
        },
        "constructInfo": {
          "fqn": "cdktf.TerraformStack",
          "version": "0.20.3",
        },
        "id": "root",
        "path": "root",
      },
    },
    "constructInfo": {
      "fqn": "cdktf.App",
      "version": "0.20.3",
    },
    "id": "App",
    "path": "",
  },
  "version": "tree-0.1",
}
`;
